### ISSUE TITLE
Neptune 3 Pro reboots and resets EEPROM after leveling bed with LCD.

### ISSUE THREAD

**Issue by L0st-Moth-of-Indigo-Swamp** [vehystrix/Elegoo-Neptune-marlin#1]
Every time I level the bed from the LCD with this firmware after I set the Z offset and exit my printer reboots and all my settings are deleted. I reverted to the official software (V 1.4.2 - 1.x.5.1b) and found the same issue. I also reverted the screen software to the official release (1.4.2). It wasn't until I switched to the V1.x.5.1c and screen firmware 1.4.3 branch that the issue seems to have disappeared. I'm not sure what the issue was but it may be that the official software (screen or board) was broken when Hummtaro originally forked it.

**Comment by L0st-Moth-of-Indigo-Swamp** [comment]
I tested the printer with the official screen firmware (1.4.3) and your custom firmware and I got the restart bug. It seems its an issue with the board firmware and not the screen.

**Comment by waffqle** [comment]
I'm encountering the same issue. ABL seems to run fine, then the printer reboots and all settings are lost. 
I'm running the latest LCD firmware from Hummtarro (dated 5/31/2024). Here: https://github.com/Hummtaro/Elegoo-Neptune-screen/releases

If there is anything testing I can run or whatnot to assist in debugging (or steps I can take myself) let me know.

Thanks!

**Comment by L0st-Moth-of-Indigo-Swamp** [comment]
I think the issue is that the official elegoo 1.x.5.1-b branch is broken, which is what hummtaro forked from and then vehystrix forked from it.

**Comment by waffqle** [comment]
For what it's worth, I rolled back to the last version Hummtaro released, based on 2.1.2.3 and it seems to have corrected the issue. I'm going through the code to see if I can ID the root of it, but I'm still new to this and it may take a bit.

**Comment by L0st-Moth-of-Indigo-Swamp** [comment]
good luck

**Comment by vehystrix** [comment]
Did you guys ever ID the issue?
I ran into a few resets myself, but completely resetting the EEPROM fixed it for me, it hasn't crashed since december
Complete reset by running `M502` followed by `M500`

Also I just updated to the latest marlin bugfix branch, and after flashing and leveling the bed it hasn't crashed (which would be where it crashed last time), so hopefully it's fixed now.

I won't be spending a lot of time figuring out why this is/was happening because it happens so rarely and only when leveling the bed, which is something you just do once and then forget about.

What I did find out when I briefly looked at it, is that the crash was happening around here https://github.com/vehystrix/Elegoo-Neptune-marlin/blob/8dc19d87aef7d2a75619000c76f6a33472b52ae4/Marlin/src/HAL/STM32/eeprom_wired.cpp#L54-L57

**Comment by awilliamwright** [comment]
> For what it's worth, I rolled back to the last version Hummtaro released, based on 2.1.2.3 and it seems to have corrected the issue. I'm going through the code to see if I can ID the root of it, but I'm still new to this and it may take a bit.

I was having this happen to me before I'd done anything to the firmware, just using the last update from Elegoo on my Neptune 3 Plus. It wouldn't happen every time, so I don't know what the x-factor was that caused it. But that was only one of many awful problems I was having with that firmware. 

Now that I've installed "2.1.2.3 build 1," which I believe is the version Waffqle is referring to, I'm now having it happen every single time I try to auto level the bed. 

HOWEVER! I discovered that if I let it auto-level the bed, then right after it finishes, I TURN OFF the printer, it will save the bed leveling information. Then I can go in and tune the z-offset, and it won't crash.

**Comment by vehystrix** [comment]
So you're saying the stock firmware also had this issue for you?
This means it's not any changes I merged into it that caused it, which is a good thing, I guess?

Note that the trick you describe won't work with the firmware builds I made, because it doesn't save directly after leveling anymore, only when you exit the leveling menu.

If you're up to it, and you've got it attached to a computer via USB, you could try the prerelease built I made which dumps a bunch of info to the serial console on every settings save. Then if it crashes we can try to see if we can detect a pattern at what point it's crashing.

**Comment by awilliamwright** [comment]
Once I have a break in my printing schedule, I can give this a go. Just to be sure, which one exactly is the prerelease you want me to test with?

**Comment by vehystrix** [comment]
Here's the release:
https://github.com/vehystrix/Elegoo-Neptune-marlin/releases/tag/N3P-bugfix-2.1.x-20250516

**Comment by j19801110** [comment]
So I rolled back all the way to https://github.com/dderg/Marlin-Neptune-3-Pro, along with LCD firmware 1.5b.

I spent a day printing different things, interacting with LCD, etc. I could not crash the printer. By this point I would have had 3-4 crashes with the latest builds.

This evening I upgrade the screen firmware to Elegoo's latest, to see if it's what causing the crashes. We shall see.

If not, then I think we can assume, there is something that Elegoo did in 1.x.5.1b that's causing some sort of issue. So starting out again from https://github.com/dderg/Marlin-Neptune-3-Pro, then merging all the latest Marlin changes back, but leaving Elegoo's mess out, might help.

However, I'll just say that even https://github.com/dderg/Marlin-Neptune-3-Pro brings noticeable benefits with regards to input shaping and maxing out print speeds and accelerations. I'm not sure I could go any faster with the stock print head. So might just stay with this build if the crashes are no longer happening. The last 5% speed increase isn't really worth the hassle. But of course with an upgraded hotend, that might be different.

**Comment by vehystrix** [comment]
Did you go back and grab the build on the github page, or did you build it yourself with their latest source?
Because they didn't do anything even remotely related to EEPROM read/write code. So the only changes between those builds from 2023 and now would be any changes in the PlatformIO Arduino library code that handles the EEPROM, and I don't have an easy way to roll those back.
If possible could you re-run your tests with the old version source but built today?

**Comment by j19801110** [comment]
I just grabbed the release that was already built at the time. Why do you think re-building it would make things different? You mean pull the latest Marlin in with all the new EEPROM changes? If that is the issue here, then there's no way forward really... 

I won't have time before I go away so you might need to wait a few weeks. I do want to rebuild the firmware, as he disabled MeatPack in his firmware build, which is a huge help when transferring stuff to the SD card from OctoPrint. :-/ But might need to wait until I'm back from holidays in 1.5 weeks.

**Comment by vehystrix** [comment]
So I did a comparison a while ago between the original Elegoo sources and the ones I have now, and none of the EEPROM code has changed between then and now.

The actual handling of EEPROM (and of all of the boards features) is done by Arduino libraries provided by the build system (PlatformIO). So while I can rule out any code changes in Marlin causing these issues, I can't rule out any changes in the Arduino libraries between when the old builds were made and now.

If the old sources built today give the same crashes as we see now, then the issue lies in changes to the libraries used to handle EEPROM. I'll have to figure out a way to track down the old library code and build with those if that's the case.

**Comment by j19801110** [comment]
Right, got ya! I didn't understand why you wanted a re-build, but that makes perfect sense. 

Ok, no dramas. I'll have to do one to enable the features I want anyway. If the crashes start happening again, then we found the culprit. Then I'm sure we can contact the team that wrote those libraries and ask for the old code... me thinking naively...

**Comment by corssoba** [comment]
Hello! I'm another person with the same problem. When updating to your latest version, it starts to reset the EPROM almost once or twice per print. Playing around with chatgtp and passing it the console log indicated that the M500 was dangerously close to its memory capacity limit. I don't know if this helps, but doing the same thing with the original Elegoo 1.05 version, it's true that the M500 stayed further away from the limit. I don't have any reboots with that version. 

What can I do to help?
How can I be more specific? 
Can you give me a quick tutorial on what to do to send you a log?
Thank you for your versions. The truth is that I find them much more pleasant to use (active screen with Octooprint) (improvements in printing, etc.).

**Comment by vehystrix** [comment]
See #3
